Userland/Libraries/LibGfx/ImageFormats/JBIG2Loader.cpp

```diff
@@ -443,6 +492,46 @@ static ErrorOr<void> warn_about_multiple_pages(JBIG2LoadingContext& context)
     return {};
 }
 
+// 6.2.2 Input parameters
+struct GenericRegionDecodingInputParameters {
+    bool is_modified_modified_read; // "MMR" in spec.
+    u32 region_width;               // "GBW" in spec.
+    u32 region_height;              // "GBH" in spec.
+    u8 gb_template;
+    bool is_typical_prediction_used;                 // "TPGDON" in spec.
+    bool is_extended_reference_template_used;        // "EXTTEMPLATE" in spec.
+    Optional<NonnullOwnPtr<BitBuffer>> skip_pattern; // "USESKIP", "SKIP" in spec.
+
+    struct AdaptiveTemplatePixel {
+        i8 x, y;
+    };
+    AdaptiveTemplatePixel adaptive_template_pixels[12]; // "GBATX" / "GBATY" in spec.
+    // FIXME: GBCOLS, GBCOMBOP, COLEXTFLAG
+};
+
+// 6.2 Generic region decoding procedure
+static ErrorOr<NonnullOwnPtr<BitBuffer>> generic_region_decoding_procedure(GenericRegionDecodingInputParameters const& inputs, ReadonlyBytes data)
+{
+    if (inputs.is_modified_modified_read) {
+        // 6.2.6 Decoding using MMR coding
+        auto buffer = TRY(CCITT::decode_ccitt_group4(data, inputs.region_width, inputs.region_height));
+        auto result = TRY(BitBuffer::create(inputs.region_width, inputs.region_height));
+        size_t bytes_per_row = ceil_div(inputs.region_width, 8);
+        if (buffer.size() != bytes_per_row * inputs.region_height)
+            return Error::from_string_literal("JBIG2ImageDecoderPlugin: Decoded MMR data has wrong size");
+
+        // FIXME: Could probably just copy the ByteBuffer directly into the BitBuffer's internal ByteBuffer instead.
```

Comment:
I'd prefer to do it later: Once I implement support for combination operators, it's possible that that's going to go slow, and maybe it could go faster by making BitBuffer operate on 32-bit words and doing ands and ors and whatnot 32 at a time instead of 8 at a time (or maybe even 64-bit). So I don't want to marry BitBuffer to an internal ByteBuffer to tightly yet. (It's also possible that it's fine as-is, and then I'll change this here instead.)